packages/theme/src/cli/utilities/theme-environment/html.ts

```diff
@@ -25,29 +25,46 @@ export function getHtmlHandler(theme: Theme, ctx: DevServerContext) {
       replaceTemplates: getInMemoryTemplates(ctx, browserPathname, getCookie(event, 'localization')?.toLowerCase()),
     })
       .then(async (response) => {
-        logRequestLine(event, response)
-
-        let html = await patchRenderingResponse(ctx, event, response)
-
-        assertThemeId(response, html, String(theme.id))
-
-        if (ctx.options.liveReload !== 'off') {
-          html = injectHotReloadScript(html)
+        if (response.status >= 400 && response.status < 500) {
+          // We have tried to render a route that can't be handled by SFR.
+          // Ideally, this should be caught by `canProxyRequest` in `proxy.ts`,
+          // but we can't be certain for all cases (e.g. an arbitrary app's route).
+          // Fallback to proxying to see if that works:
+
+          outputDebug(`Render failed for ${event.path} with ${response.status}, trying proxy...`)
+
+          // eslint-disable-next-line promise/no-nesting
+          const proxyResponse = await proxyStorefrontRequest(event, ctx).catch(
+            (error: H3Error) => new Response(null, {status: error.statusCode ?? 502}),
+          )
+
+          if (proxyResponse.status < 400) {
+            outputDebug(`Proxy status: ${proxyResponse.status}. Returning proxy response.`)
+            logRequestLine(event, proxyResponse)
+            return proxyResponse
+          } else {
+            outputDebug(`Proxy status: ${proxyResponse.status}. Returning render error.`)
+          }
         }
 
-        return html
+        logRequestLine(event, response)
+
+        return patchRenderingResponse(ctx, response, (body) => {
+          assertThemeId(response, body, String(theme.id))
+          return ctx.options.liveReload === 'off' ? body : injectHotReloadScript(body)
+        })
       })
       .catch(async (error: H3Error<{requestId?: string; url?: string}>) => {
-        let headline = `Failed to render storefront with status ${error.statusCode} (${error.statusMessage}).`
+        const status = error.statusCode ?? 502
+        const statusText = error.statusMessage ?? 'Bad Gateway'
+
+        let headline = `Failed to render storefront with status ${status} (${statusText}).`
```

Comment:
We're getting a 404 here when running the development server via `shopify app dev`. I've raised an issue for supporting the final fix for [this](https://github.com/Shopify/developer-tools-team/issues/543).

While we don't have that fixed, can we prevent the warning from being rendered here when the `app_block_id` query parameter is passed? (Alternatively, instead of preventing it, we could still render the warning here in the verbose output in that context.)

Reply:
Precisely, I believe that's very nice to have, but we may also apply that fix in another PR, as it transcends a bit the scope of this one.